### PR TITLE
feat: new admin user punches

### PR DESCRIPTION
**Pull request by bedreski**
WIP 
Solves: #392 
This feature is focused in the Punches tab, inside new admin user area. 

The filter feature is not working properly. We believe that it could be the date format between the form field and the created_at attribute, but I think I'm missing another detail. It would be very helpful if the review point this up. Thanks.

### CODE REVIEW

File: app/models/punch.rb [hunk 1]
```diff
@@ -21,6 +21,7 @@ class Punch < ApplicationRecord
     last_month = current_month.prev_month + 1.day
     where from: last_month..current_month
   }
+  scope :filter_by_date, -> (user_id, from, to) { where('user_id = ? AND created_at >= ? AND created_at <= ?', user_id, from.to_datetime, to.to_datetime) }
```
Comment (adbatista) [inline app/models/punch.rb:24]: specs missing

File: app/decorators/punch_decorator.rb [hunk 2]
```diff
@@ -33,6 +33,10 @@ def delta_in_hours
     object.delta / 1.hour
   end
 
+  def extra_hour
+    I18n.t(model.extra_hour.to_s)
+  end
```
Comment (adbatista) [inline app/decorators/punch_decorator.rb:38]: specs missing

File: app/controllers/new_admin/users_controller.rb [hunk 3]
```diff
@@ -9,6 +9,11 @@ def show
       @user_allocations = Allocation.where(user_id: params[:id]).decorate
       @performance_evaluations = Evaluation.joins(:questionnaire).where(questionnaires: { kind: :performance }, evaluated_id: params[:id]).order(created_at: :desc).decorate
       @english_evaluations = Evaluation.joins(:questionnaire).where(questionnaires: { kind: :english }, evaluated_id: params[:id]).order(created_at: :desc).decorate
+      if params[:from].nil? && params[:to].nil?
+        @punches = Punch.where(user_id: params[:id]).decorate
+      else
+        @punches = Punch.all.filter_by_date(params[:id], params[:from], params[:to]).decorate
+      end
```
Comment (adbatista) [inline app/controllers/new_admin/users_controller.rb:16]: specs missing

File: spec/features/new_admin/user_punches_spec.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    it 'shows page with user punches', js: true do 
+      visit "/new_admin/users/#{user.id}"
+      click_button('Punches')
+      expect(page).to have_content("PROJETO")
+      expect(page).to have_content(punch.project.name)
+      expect(page).to have_content("HORÁRIO INICIAL")
+      expect(page).to have_content("ATÉ")
+      expect(page).to have_content("DELTA")
+      expect(page).to have_content("HORA EXTRA")
```
Comment (adbatista) [inline spec/features/new_admin/user_punches_spec.rb:34]: add composable matches pls

File: spec/features/new_admin/user_punches_spec.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      visit "/new_admin/users/#{user.id}"
+      click_button('Punches')
+      expect(page).to have_content("Filtro")
+      expect(page).to have_content("INTERVALO")
+      expect(page).to have_field("from")
+      expect(page).to have_field("to")
+      expect(page).to have_button("Filtrar")
+      expect(page).to have_link("Limpar Filtro")
+      expect(page).to have_link("Baixar XLS")
+      expect(page).to have_link("Todos os Punches")
```
Comment (adbatista) [inline spec/features/new_admin/user_punches_spec.rb:47]: add composable matches pls

File: app/controllers/new_admin/users_controller.rb [hunk 6]
```diff
@@ -9,6 +9,11 @@ def show
       @user_allocations = Allocation.where(user_id: params[:id]).decorate
       @performance_evaluations = Evaluation.joins(:questionnaire).merge(Questionnaire.performance).where(evaluated_id: params[:id]).order(created_at: :desc).decorate
       @english_evaluations = Evaluation.joins(:questionnaire).merge(Questionnaire.english).where(evaluated_id: params[:id]).order(created_at: :desc).decorate
+      if params[:from].nil? and params[:to].nil?
+        @punches = Punch.where(user_id: params[:id]).decorate
+      else
+        @punches = Punch.all.filter_by_date(params[:id], params[:from], params[:to]).decorate
```
Comment (adbatista) [inline app/controllers/new_admin/users_controller.rb:15]: you don't need `.all` here

File: spec/features/new_admin/user_punches_spec.rb (hunk 5 above)
Comment (adbatista) [reply spec/features/new_admin/user_punches_spec.rb:47]: please add composable matches

File: spec/features/new_admin/user_punches_spec.rb (hunk 4 above)
Comment (adbatista) [reply spec/features/new_admin/user_punches_spec.rb:34]: same here pls

File: spec/decorators/punch_decorator_spec.rb [hunk 7]
```diff
@@ -2,4 +2,21 @@
...
+      expect(punch_decorator.extra_hour).to eq("#{I18n.t(false)}")
+    end 
+  end 
+
+  describe "#extra_hour as 'yes'" do 
+    let(:punch) { create :punch, :is_extra_hour }
+    let(:punch_decorator) { PunchDecorator.new(punch) }
+
+    it "translates the bool value into 'yes'" do 
+      expect(punch_decorator.extra_hour).to eq("#{I18n.t(true)}")
```
Comment (adbatista) [inline spec/decorators/punch_decorator_spec.rb:19]: here you should check the text ` expect(punch_decorator.extra_hour).to eq("yes")`

if for some reason, someone changes the translation text from yes to another text this spec should break, if you check against the text from the translation `I18n.t(true)` will give an false positive

Comment (bedreski) [reply spec/decorators/punch_decorator_spec.rb:19]: Oh, that's so true!

**Pull request merged**